Objects/unionobject.c

```diff
@@ -0,0 +1,435 @@
+}
+
+static Py_hash_t
+union_hash(PyObject *self)
+{
+    unionobject *alias = (unionobject *)self;
+    Py_hash_t h1 = PyObject_Hash(alias->args);
+    if (h1 == -1) {
+        return -1;
+    }
+    return h1;
+}
+
+static int
+union_traverse(PyObject *self, visitproc visit, void *arg)
+{
+    unionobject *alias = (unionobject *)self;
+    Py_VISIT(alias->args);
+    return 0;
+}
+
+static PyMemberDef union_members[] = {
+    {"__args__", T_OBJECT, offsetof(unionobject, args), READONLY},
+    {0}
+};
+
+static PyObject *
+union_getattro(PyObject *self, PyObject *name)
+{
+    return PyObject_GenericGetAttr(self, name);
+}
+
+static PyObject *
+union_instancecheck(PyObject *self, PyObject *instance)
+{
+    unionobject *alias = (unionobject *) self;
+    Py_ssize_t nargs = PyTuple_GET_SIZE(alias->args);
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+        if (PyType_Check(arg)) {
+            if (PyObject_IsInstance(instance, arg) != 0)
+            {
+                return self;
+            }
+        }
+    }
+    return instance;
+}
+
+static PyObject *
+union_subclasscheck(PyObject *self, PyObject *instance)
+{
+    unionobject *alias = (unionobject *) self;
+    Py_ssize_t nargs = PyTuple_GET_SIZE(alias->args);
+    for (Py_ssize_t iarg = 0; iarg < nargs; iarg++) {
+        PyObject *arg = PyTuple_GET_ITEM(alias->args, iarg);
+        if (PyType_Check(arg)) {
+            if (PyType_IsSubtype((PyTypeObject *)instance, (PyTypeObject *)arg) != 0)
+            {
+                return self;
+            }
+        }
+    }
+    return instance;
+}
+
+static int
+is_typing_name(PyObject *obj, char *name)
+{
+    PyTypeObject *type = Py_TYPE(obj);
+    if (strcmp(type->tp_name, name) != 0) {
+        return 0;
+    }
+    PyObject *module = PyObject_GetAttrString((PyObject *)type, "__module__");
+    if (module == NULL) {
+        return -1;
+    }
+    int res = PyUnicode_Check(module)
+        && _PyUnicode_EqualToASCIIString(module, "typing");
+    Py_DECREF(module);
+    return res;
+}
+
+static PyMethodDef union_methods[] = {
+    {"__instancecheck__", union_instancecheck, METH_O},
+    {"__subclasscheck__", union_subclasscheck, METH_O},
+    {0}};
+
+static PyObject *
+union_richcompare(PyObject *a, PyObject *b, int op)
+{
+    if (op != Py_EQ && op != Py_NE) {
+        PyObject *result = Py_NotImplemented;
+        Py_INCREF(result);
+        return result;
+    }
+
+    PyObject* b_set = NULL;
+    unionobject *aa = (unionobject *)a;
+    PyObject* a_set = PySet_New(aa->args);
+
+    if (is_typing_name(b, "_UnionGenericAlias")) {
+        PyObject* b_args = PyObject_GetAttrString(b, "__args__");
+        b_set = PySet_New(b_args);
+    }
+
+    PyTypeObject *type = Py_TYPE(b);
+    if (type == &Py_UnionType) {
+        unionobject *bb = (unionobject *)b;
+        b_set = PySet_New(bb->args);
+    }
+
+    if (b_set == NULL || a_set == NULL) {
```

Comment:
@pablogsal  they are both checked here.

Reply:
You should check them individually (or at least in a way that you don't try to construct `b_set` if `a_set` fails). This is because if `a_set` fails it will set an exception and if-then `b_set` fails it will override it or even crash (you cannot call Python code with the error indicator set).

Reply:
Ah, thanks for the explanation - I will update this check :)